misk/src/main/kotlin/misk/web/interceptors/RequestLoggingInterceptor.kt

```diff
@@ -24,88 +25,123 @@ private val logger = getLogger<RequestLoggingInterceptor>()
+      }
+      require(logRequestResponse.errorBodySampling in 0.0..1.0) {
         "${action.name} @LogRequestResponse sampling must be in the range (0.0, 1.0]"
       }
 
       return RequestLoggingInterceptor(
         action,
-        logRequestResponse.sampling,
-        logRequestResponse.includeBody,
         caller,
         ticker,
         random,
+        logRateLimiter,
+        logRequestResponse.ratePerSecond,
+        logRequestResponse.errorRatePerSecond,
+        logRequestResponse.bodySampling,
+        logRequestResponse.errorBodySampling,
         bodyCapture
       )
     }
   }
 
   override fun intercept(chain: NetworkChain) {
-    val randomDouble = random.current().nextDouble()
-    if (randomDouble >= sampling) {
-      return chain.proceed(chain.httpCall)
-    }
-
-    val principal = caller.get()?.principal ?: "unknown"
-
     bodyCapture.clear()
 
     val stopwatch = Stopwatch.createStarted(ticker)
 
+    var error: Throwable? = null
     try {
-      val result = chain.proceed(chain.httpCall)
+      chain.proceed(chain.httpCall)
       stopwatch.stop()
+    } catch (e: Throwable) {
+      error = e
+    }
 
-      logRequestResponse(principal, stopwatch, chain.httpCall.statusCode)
+    try {
+      maybeLog(chain.httpCall, stopwatch, error)
+    } catch (e: Throwable) {
+      logger.error(e) { "Unexpected error while logging request" }
+    }
 
-      return result
-    } catch (t: Throwable) {
-      logRequestResponse(principal, stopwatch, null)
-      throw t
+    if (error != null) {
+      throw error
     }
   }
 
-  fun logRequestResponse(principal: String, stopwatch: Stopwatch, statusCode: Int?) {
+  fun maybeLog(httpCall: HttpCall, stopwatch: Stopwatch, error: Throwable?) {
+    val principal = caller.get()?.principal ?: "unknown"
+
     val builder = StringBuilder()
     builder.append("${action.name} principal=$principal time=$stopwatch")
 
-    if (statusCode == null) {
+    val statusCode = httpCall.statusCode
+    if (error != null) {
       builder.append(" failed")
     } else {
       builder.append(" code=${statusCode}")
     }
 
-    if (includeBody) {
-      val requestResponseBody = bodyCapture.get()
-      requestResponseBody?.let {
-        requestResponseBody.request?.let {
-          builder.append(" request=${requestResponseBody.request}")
-        }
-        requestResponseBody.response?.let {
-          builder.append(" response=${requestResponseBody.response}")
-        }
+    val isError = statusCode > 299 || error != null
+
+    val rateLimit = if (isError) errorRatePerSecond else ratePerSecond
+    val loggingBucketId = LogBucketId(actionClass = action.name, isError = isError)
+    if (!logRateLimiter.tryAcquire(loggingBucketId, rateLimit)) {
+      return
+    }
+
+    val sampling = if (isError) errorBodySampling else bodySampling
+    val randomDouble = random.current().nextDouble()
+    if (randomDouble >= sampling) {
+      logger.info(
```

Comment:
since you are building the string, it would be nicer to keep going here rather than short circuiting. Negate the conditional to `if (randomDouble < sampling)` and add the body

Reply:
Ah yeah much cleaner